### ISSUE TITLE
Websocket TCP close behaviour

### ISSUE THREAD

**Issue by dmpriso** [dotnet/corefx#33339]
We're running a .NET Core (Runtime 2.1.5) Websocket server (with `Net.HttpListener`) on Linux (Ubuntu 18.04)

When accessing the websocket server from a `boost::beast` C++ client, we cannot shut the connection down in under 30 seconds. 

Using network analyzers, we were able to find the problem.
When the client sends the close packet (with `WebSocketCloseStatus.NormalClosure`), the server also replies with a close packet correctly, because our server code reacts with `CloseOutputAsync`. However, contrary to what the RFC suggests (if I read [section 7.1.1](https://tools.ietf.org/html/rfc6455#section-7.1.1) correctly), the dotnet core websocket server implementation does not initiate closure of the underlying TCP connection (by sending a `FIN` packet). Even worse, it sends a `PING` 30 seconds later on the already-closed websocket connection (which makes the `beast` angry and close the connection forcefully).

The same problem happens when connecting a websocket from google chrome btw.

If using a C# client websocket, the behaviour is different in that the client initiates the TCP connection teardown and so the websocket shutdown finishes quickly on both sides.

BTW it's not even possible to forcefully close the underlying connection using `Abort` or `Dispose` after calling `CloseOutputAsync`. Using `CloseAsync` instead also doesn't cure the problem.

**Comment by davidsh** [comment]
@stephentoub

**Comment by karelz** [comment]
Can you please check also Kestrel-based WebSocket server? (`HttpListener` is obsolete)
If you can post minimal repro to reproduce the above, that would be very helpful. Thanks!

**Comment by dmpriso** [comment]
Minimal example C# server code:

```
using System;
using System.Net;
using System.Net.WebSockets;
using System.Threading;
using System.Threading.Tasks;

namespace WebsocketTest
{
    class Program
    {
        static async Task HandleContext(HttpListenerContext ctx)
        {
            var ws = (await ctx.AcceptWebSocketAsync(null)).WebSocket;
            var readbuf = new ArraySegment<byte>(new Byte[65535]);
            while(true)
            {
                var msg = await ws.ReceiveAsync(readbuf, CancellationToken.None);
                if (msg.MessageType == System.Net.WebSockets.WebSocketMessageType.Close)
                {
                    await ws.CloseOutputAsync(WebSocketCloseStatus.NormalClosure, "", CancellationToken.None);
                    return;
                }
                else
                {
                    var payload =  new ArraySegment<byte>(readbuf.Array, readbuf.Offset, msg.Count);
                    await ws.SendAsync(payload, msg.MessageType, true, CancellationToken.None);
                }
            }
        }

        static async Task RunServer(CancellationToken cancel)
        {
            var url = "http://localhost:3001/";

            var listener = new HttpListener();
            listener.Prefixes.Add(url);
            listener.Start();

            bool cancelled = false;
            cancel.Register(() => { cancelled = true; listener.Stop(); });

            while(true)
            {
                HttpListenerContext context;
                try
                {
                    context = await listener.GetContextAsync();
                }
                catch(Exception)
                {
                    if (cancelled)
                        break;
                    throw;
                }
                await HandleContext(context);
            }
        }

        static void Main(string[] args)
        {
            var source = new CancellationTokenSource();
            var serverTask = RunServer(source.Token);
            serverTask.ContinueWith(t => {
                Console.WriteLine($"Server failed with {t.Exception}");
            }, TaskContinuationOptions.OnlyOnFaulted);

            Console.WriteLine("Press ENTER to stop");
            Console.ReadLine();
            source.Cancel();
            serverTask.Wait();
        }
    }
}
```

Node.js test client:
```
var WebSocketClient = require('websocket').client;

var client = new WebSocketClient();
client.on('connectFailed', function(error) {
    console.log(`Connect error ${error.toString()}`);
});

client.on('connect', function(connection)
{
    console.log('Connected.');
    connection.on('error', function(error) {
        console.log(`Connection error ${error.toString()}`);
    });
    connection.on('close', function() {
        console.log(`Connection closed`);
    });
    connection.on('message', function(message) {
        console.log(`Received message of type ${message.type}`);
        if (message.type === 'utf8') {
            console.log(`Received: ${message.utf8Data}`);
        };
        
        console.log("Closing connection");
        connection.close();
    });

    var testMessage = 'The quick brown fox jumps over the lazy programmer\'s dog';
    console.log(`Sending test message: ${testMessage}`);
    connection.sendUTF(testMessage);
});

client.connect('ws://localhost:3001/');
```

package.json:
```
{
  "name": "websockettest",
  "version": "1.0.0",
  "description": "",
  "main": "testclient.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "author": "",
  "license": "ISC",
  "dependencies": {
    "websocket": "^1.0.28"
  }
}
```

Wireshark trace:
[websocket-trace.pcapng.zip](https://github.com/dotnet/corefx/files/2575106/websocket-trace.pcapng.zip)

**Comment by dmpriso** [comment]
Kestrel behaves correctly, tested with `boost::asio` and `node`

**Comment by stephentoub** [comment]
The issue is that the ManagedWebSocket implementation isn't closing the web socket after the close handshake has completed when the other end of the connection initiates the close.  I've removed the os-linux label, as this also affects ClientWebSocket... it's just not as obvious there because generally after closing, you then dispose of the ClientWebSocket, which achieves the same result.  I put up a fix in #33473.
